design/automatic-network-configuration.md

```diff
@@ -0,0 +1,355 @@
+# Automatic network configuration
+
+## Status
+provisional
+
+## Summary
+This proposal is to expand the scope of Metal³ to include an API to manage
+physical network devices.
+
+## Motivation
+Metal³ follows the paradigm of Kubernetes Native Infrastructure (KNI), which is
+an approach to use Kubernetes to manage underlying infrastructure.
+Managing the configuration of some physical network devices is closely related
+to managing physical hosts.
+
+As bare metal hosts are provisioned or later repurposed, there may be
+corresponding physical network changes that must be made, such as reconfiguring
+a ToR switch port. If the provisioning of the physical host is managed through
+a Kubernetes API, it would be convenient to be able to reconfigure related
+network devices using a similar API.
+
+### Goals
+- Define a Kubernetes API for configuring network switches.
+- Automatically configure the network infrastructure for a host when adding it
+  to a cluster.
+
+### Non-Goals
+- Add new Switch API to BMO.
+
+## Proposal
+This document proposes to add a new mechanic to automatically perform physical
+network device configuration before provisioning a BareMetalHost.
+It contains 4 new objects (Switch, SwitchPort, Interface and Network object), 3
+new controllers (Switch controller, Interface controller and Network
+controller) and related changes to current code including CAPM3, BareMetal
+Operator, Metal3Cluster and BareMetalHost.
+
+### User Stories
+
+#### Story 1
+As a consumer of Metal³, when adding a machine resource to the cluster, it can
+automatically put the adapted BMH into the correct Switch's VLAN.
+
+#### Story 2
+As a user/admin a SmartNIC that is hosted in a BMH needs to be provisioned,
+pre-configured, and then capable of in operation management.
+
+#### Story 3
+As an admin, when creating a new machine, it is possible to automate configure
+LAG/mLAG.
+
+#### Story 4
+As an admin, when creating a cluster, it is possible to configure ACL settings
+in a switch for each bare metal server according to its functionality.
+The ACL setting:
+- Direction: ingress or egress
+- Protocol: SSH, HTTP, …all
+- Action: allow, deny
+- Port status: up, down (status of the switch port)
```

Comment:
Because we are not familiar with this, many situations are not considered. If you have a good idea about this, you can help us improve this story, thank you very much.